code/modules/mod/mod_modules.dm

```diff
@@ -0,0 +1,607 @@
+/obj/item/mod/module
+	name = "MOD module"
+	icon_state = "module"
+	/// If it can be removed
+	var/removable = TRUE
+	/// If it's passive, active or usable
+	var/module_type = MODULE_PASSIVE
+	/// Is the module active
+	var/active = FALSE
+	/// How much space it takes up in the MOD
+	var/complexity = 0
+	/// Power use when idle
+	var/idle_power_cost = 0
+	/// Power use when active
+	var/active_power_cost = 0
+	/// Power use when used
+	var/use_power_cost = 0
+	/// ID used by their TGUI
+	var/tgui_id = null
+	/// Can we be configured
+	var/configurable = FALSE
+	/// Linked MODsuit
+	var/obj/item/mod/control/mod
+	/// If we're an active module, what item are we?
+	var/obj/item/device
+	/// Overlay given to the user when the module is inactive
+	var/overlay_state_inactive
+	/// Overlay given to the user when the module is active
+	var/overlay_state_active
+	/// What modules are we incompatible with?
+	var/list/incompatible_modules = list()
+	/// Cooldown after use
+	var/cooldown_time = 0
+	/// Timer for the cooldown
+	COOLDOWN_DECLARE(cooldown_timer)
+
+/obj/item/mod/module/Initialize()
+	. = ..()
+	if(module_type != MODULE_ACTIVE)
+		return
+	if(ispath(device))
+		device = new device(src)
+		ADD_TRAIT(device, TRAIT_NODROP, MOD_TRAIT)
+		RegisterSignal(device, COMSIG_PARENT_PREQDELETED, .proc/on_device_deletion)
+		RegisterSignal(src, COMSIG_ATOM_EXITED, .proc/on_exit)
+
+/obj/item/mod/module/Destroy()
+	if(mod)
+		mod.uninstall(src)
```

Comment:
Haven't checked, but this is guaranteed both to delete itself AND not perform any side effects, correct?